core/src/main/java/com/datastax/oss/driver/internal/core/auth/PlainTextAuthProviderBase.java

```diff
@@ -0,0 +1,166 @@
+   *
+   * <p>This is invoked every time the driver opens a new connection.
+   */
+  @NonNull
+  protected abstract Credentials getCredentials();
+
+  @NonNull
+  @Override
+  public Authenticator newAuthenticator(
+      @NonNull EndPoint endPoint, @NonNull String serverAuthenticator) {
+    return new PlainTextAuthenticator(getCredentials());
+  }
+
+  @Override
+  public void onMissingChallenge(@NonNull EndPoint endPoint) {
+    LOG.warn(
+        "[{}] {} did not send an authentication challenge; "
+            + "This is suspicious because the driver expects authentication",
+        logPrefix,
+        endPoint);
+  }
+
+  @Override
+  public void close() {
+    // nothing to do
+  }
+
+  protected static class Credentials {
+
+    private final char[] username;
+    private final char[] password;
+
+    public Credentials(@NonNull char[] username, @NonNull char[] password) {
+      this.username = Objects.requireNonNull(username);
+      this.password = Objects.requireNonNull(password);
+    }
+
+    @NonNull
+    public char[] getUsername() {
+      return username;
+    }
+
+    @NonNull
+    public char[] getPassword() {
+      return password;
+    }
+
+    /** Clears the credentials from memory when they're no longer needed. */
+    protected void clear() {
+      // Note: this is a bit irrelevant with the built-in provider, because the config already
+      // caches the credentials in memory. But it might be useful for a custom implementation that
+      // retrieves the credentials from a different source.
+      Arrays.fill(getUsername(), (char) 0);
+      Arrays.fill(getPassword(), (char) 0);
+    }
+  }
+
+  protected static class PlainTextAuthenticator implements SyncAuthenticator {
+
+    private final ByteBuffer initialToken;
+
+    protected PlainTextAuthenticator(@NonNull Credentials credentials) {
+      Objects.requireNonNull(credentials);
+      ByteBuffer usernameBytes = toUtf8Bytes(credentials.getUsername());
+      ByteBuffer passwordBytes = toUtf8Bytes(credentials.getPassword());
+      credentials.clear();
+
+      this.initialToken =
+          ByteBuffer.allocate(usernameBytes.remaining() + passwordBytes.remaining() + 2);
+      initialToken.put((byte) 0);
+      initialToken.put(usernameBytes);
+      initialToken.put((byte) 0);
+      initialToken.put(passwordBytes);
+      initialToken.flip();
+
+      // Clear temporary buffers
+      usernameBytes.rewind();
+      while (usernameBytes.remaining() > 0) {
+        usernameBytes.put((byte) 0);
+      }
+      passwordBytes.rewind();
+      while (passwordBytes.remaining() > 0) {
+        passwordBytes.put((byte) 0);
+      }
+    }
+
+    private ByteBuffer toUtf8Bytes(char[] charArray) {
+      CharBuffer charBuffer = CharBuffer.wrap(charArray);
+      return Charsets.UTF_8.encode(charBuffer);
+    }
+
+    @Override
+    @Nullable
+    public ByteBuffer initialResponseSync() {
+      return initialToken.duplicate();
```

Comment:
Continuing with the idea that sensitive information should be kept in memory as shortly as possible, this should probably be cleared once the protocol layer is done with it.

That's not really urgent (currently it will be GCed with the `Authenticator` / `ProtocolInitHandler`), I've logged [JAVA-2306](https://datastax-oss.atlassian.net/browse/JAVA-2306) to follow up on it.